{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This script will query the sketched k-mers against a soil metagenome bloom filter\n",
    "import subprocess\n",
    "import os\n",
    "import numpy as np\n",
    "import MinHash as MH\n",
    "from itertools import *\n",
    "from multiprocessing.dummy import Pool\n",
    "import multiprocessing\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "\n",
    "metagenome_bloom_filter = sys.argv[1]\n",
    "total_kmers_file = os.path.abspath('../data/Output/' + os.path.basename(metagenome_bloom_filter.replace(\".small\", \"\")) + '-distinctKmers.txt')\n",
    "metagenome_base_name = os.path.basename(metagenome_bloom_filter)\n",
    "\n",
    "# Get the total number of distinct kmers\n",
    "jellyfish_loc = \"/home/pi/koslickd/jellyfish-2.2.3/bin/./jellyfish\"\n",
    "\n",
    "# Get the number of unique k-mers in the metagenome\n",
    "#cmd = jellyfish_loc + \" stats \" + metagenome_bloom_filter\n",
    "#res = subprocess.check_output(cmd, shell=True)\n",
    "#num_kmers = int(res.split()[3])\n",
    "#fid = open(total_kmers_file, 'w')\n",
    "#fid.write(\"%d\" % num_kmers)\n",
    "#fid.close()\n",
    "\n",
    "\n",
    "num_threads = multiprocessing.cpu_count()\n",
    "ksize = 21  # k-mer length\n",
    "max_h = 500  # max number of hashes in sketch\n",
    "p = 0.01\n",
    "query_per_sequence_loc = os.path.abspath('QueryPerSequence/./query_per_sequence')\n",
    "#metagenome_bloom_filter = os.path.abspath('../data/MetagenomeBloom.jf')\n",
    "#metagenome_kmers_total = 916485607  # via jellyfish stats on a count bloom filter (need to streamline this)\n",
    "# Note that this number changed since I am restricting myself to the paired reads (no orphaned guys)\n",
    "#fid = open(os.path.abspath(\"../Paper/Data/MetagenomeTotalKmers.txt\"), 'r')\n",
    "fid = open(os.path.abspath(total_kmers_file), 'r')\n",
    "metagenome_kmers_total = int(fid.readlines()[0].strip())\n",
    "fid.close()\n",
    "\n",
    "# Get virus names\n",
    "file_names = list()\n",
    "fid = open(os.path.abspath('../data/Viruses/FileNames.txt'), 'r')\n",
    "for line in fid.readlines():\n",
    "\tfile_names.append(os.path.abspath(os.path.join('../data/Viruses/', os.path.basename(line.strip()))))\n",
    "fid.close()\n",
    "\n",
    "# Get all the hashes so we know the size\n",
    "base_names = [os.path.basename(item) for item in file_names]\n",
    "genome_sketches = MH.import_multiple_from_single_hdf5(os.path.abspath('../data/Viruses/AllSketches.h5'), base_names)\n",
    "\n",
    "\n",
    "# query the bloom filter\n",
    "def count_jaccard(genome_sketch, file_name, ksize, p, max_h, metagenome_kmers_total):\n",
    "\tname = os.path.basename(file_name)\n",
    "\tCMH = genome_sketch\n",
    "\tgenome_kmers_len = CMH._true_num_kmers\n",
    "\tcmd = query_per_sequence_loc + \" \" + metagenome_bloom_filter + \" \" + \\\n",
    "\t\tos.path.abspath(os.path.join('../data/Viruses/', name + \".Hash\" + str(ksize) + \"mers.fa\"))\n",
    "\tint_est = int(subprocess.check_output(cmd, shell=True))\n",
    "\tint_est -= p*int_est\n",
    "\tcontainment_est = int_est / float(max_h)\n",
    "\tcontainment_est_jaccard = genome_kmers_len * containment_est / \\\n",
    "\t\t\t\t\t\t\t(genome_kmers_len + metagenome_kmers_total - genome_kmers_len * containment_est)\n",
    "\treturn containment_est_jaccard\n",
    "\n",
    "def unwrap_make_jaccard(arg):\n",
    "\treturn count_jaccard(*arg)\n",
    "\n",
    "pool = Pool(processes=num_threads)\n",
    "jaccards = np.array(pool.map(unwrap_make_jaccard, zip(genome_sketches, file_names, repeat(ksize), repeat(p), repeat(max_h), repeat(metagenome_kmers_total))))\n",
    "\n",
    "np.savetxt(os.path.abspath('../data/Output/' + metagenome_base_name + '-Viruses.jaccards.txt'), jaccards)\n",
    "\n",
    "\n",
    "# What was the largest containment guy (If I want an estimate of the pure cardinality of the intersection,\n",
    "# then use containments.append(containment_est*max_h)\n",
    "it = 0\n",
    "containments = list()\n",
    "for genome in file_names:\n",
    "\tCMH = genome_sketches[it]\n",
    "\tgenome_kmers_len = CMH._true_num_kmers\n",
    "\tj = jaccards[it]\n",
    "\tg = genome_kmers_len\n",
    "\tm = metagenome_kmers_total\n",
    "\tcontainment_est = (j*(g+m))/float(g*(1+j))\n",
    "\tit += 1\n",
    "\tcontainments.append(containment_est)\n",
    "\n",
    "#plt.figure()\n",
    "#plt.plot(containments)\n",
    "pos = np.array(containments).argmax()\n",
    "name = os.path.basename(file_names[pos])\n",
    "fid = open(file_names[pos], 'r')\n",
    "head = fid.readline()\n",
    "fid.close()\n",
    "accession = head.split(' ')[0][1:]\n",
    "head = ' '.join(head.split(',')[0].split(' ')[1:])\n",
    "fid = open(os.path.abspath('../data/Output/' + metagenome_base_name + '-FoundOrganismName.txt'), 'w')\n",
    "fid.write(\"%s\" % head)\n",
    "fid.close()\n",
    "fid = open(os.path.abspath('../data/Output/' + metagenome_base_name + '-FoundOrganismAccession.txt'), 'w')\n",
    "fid.write(\"%s\" % accession)\n",
    "fid.close()\n",
    "fid = open(os.path.abspath('../data/Output/' + metagenome_base_name + '-FoundOrganismFileName.txt'), 'w')\n",
    "fid.write(\"%s\" % file_names[pos])\n",
    "fid.close()\n",
    "fid = open(os.path.abspath('../data/Output/' + metagenome_base_name + '-FoundOrganismContainment.txt'), 'w')\n",
    "fid.write(\"%1.4f\" % containments[pos])\n",
    "fid.close()\n",
    "fid = open(os.path.abspath('../data/Output/' + metagenome_base_name + '-FoundOrganismJaccard.txt'), 'w')\n",
    "fid.write(\"%.3e\" % jaccards[pos])\n",
    "fid.close()\n",
    "\n",
    "\n",
    "# What was the largest jaccard, don't think this is a really good way to get the guys in there\n",
    "# coverage is a better estimate of presence/absence (jaccard just returns megaviruses)\n",
    "#pos = jaccards.argmax()\n",
    "#name = os.path.basename(file_names[pos])\n",
    "#plt.figure()\n",
    "#plt.plot(jaccards)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
